diff --git a/torch/csrc/utils/python_arg_parser.h b/torch/csrc/utils/python_arg_parser.h
index 9b23af58297..57eae74bf11 100644
--- a/torch/csrc/utils/python_arg_parser.h
+++ b/torch/csrc/utils/python_arg_parser.h
@@ -840,7 +840,11 @@ inline at::Device toDevice(PyObject* obj) {
   if (THPUtils_checkLong(obj)) {
     const auto device_index = THPUtils_unpackLong(obj);
     TORCH_CHECK(device_index >= 0, "Device index must not be negative");
+#ifdef USE_CUDA
     return at::Device(DeviceType::CUDA, device_index);
+#else
+    return at::Device(DeviceType::MLU, device_index);
+#endif
   }
   const std::string& device_str = THPUtils_unpackString(obj);
   return at::Device(device_str);
